Add support for branch and path coverage

jubianchi opened the pull request:
Here comes branch and path coverage for all of you coverage lovers!

## Requirements:

To test this new feature you will need xDebug `>= 2.3`. To install it, use [pickle](https://github.com/FriendsOfPHP/pickle)

```
$ pickle install xdebug
```

## Testing this branch:

The patch is actually in its own branch since it's not finished. If you want to play with it you'll have to tweak your `composer.json`:

```json
{
	"require": {
		"atoum/atoum": "dev-branch-path-coverage as 2.1.0"	
	}	
}
```

## Use it:

Enabling branch and path coverage is as simple as appending an extra command line argument when you run your tests:

```
vendor/bin/atoum -d tests/units --enable-branch-and-path-coverage # or -ebpc
```

You test suite will run as usual but you will get extra informations on the command line output:

![atoum_coverage](https://cloud.githubusercontent.com/assets/327237/7139022/389df178-e2c4-11e4-85f3-5b0ab6022b11.png)

## Todo:

* [ ] Display branch coverage in the HTML report
* [ ] Display path coverage in the HTML report

Hywan commented:
:+1:!

jubianchi commented:
Hey guys!

What do you think of this for detailed branch/path coverage in the CLI report ?

![atoum-bpc](https://cloud.githubusercontent.com/assets/327237/7160934/71a549bc-e38c-11e4-9d67-cc0b0636b6f8.png)

Hywan commented:
@jubianchi Add commas between information, something like: `Line: %f%%, Path: %f%%, Branch: %f%%`.

jubianchi commented:
@Hywan right!

jubianchi commented:
> Display branch coverage in the HTML report
>
> Display path coverage in the HTML report

This has been done in an extension: https://github.com/atoum/reports-extension

I'm cleaning a few things and everything should be ready for merge tonight.

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/atoum/atoum/inspections/75627c2d-9f5d-4845-a405-0bc67df9af4d) was created.

jubianchi commented:
Everything should be ok now.

I fixed and updated everything and check as much as I can the reports. Everything seems ok to me but if somebody wants to double-check, here is a report generated with atoum's test suite: http://atoum.github.io/reports-extension/index.html

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/atoum/atoum/inspections/45db5db3-2f51-4a38-aeed-4b401e7eca2e) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/atoum/atoum/inspections/75627c2d-9f5d-4845-a405-0bc67df9af4d) completed: **20 updated code elements**

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/atoum/atoum/inspections/45db5db3-2f51-4a38-aeed-4b401e7eca2e) completed: **20 updated code elements**

jubianchi commented:
FYI, on my laptop when `-ebpc` is enabled, atoum's test suite takes about 14 minutes and 5 minutes without `-ebpc` so the overhead is huge.

Hywan commented:
@jubianchi: Do we now why such an overhead? Should we trace a little bit?

jubianchi commented:
@hywan it mainly comes from xdebug: producing branches ans paths informations is heavy

jubianchi commented:
I'm going to merge this PR today. If anyone has anything to say then it's the right time to do so. 

Without any feedback in the coming hours I'll click this shiny green merge button.

mageekguy commented:
Just do it, Man!
And thanks for your work, you're amazing.

ypereirareis commented:
Hurry up !! :smiley:

scrutinizer-notifier commented:
A [new inspection](https://scrutinizer-ci.com/g/atoum/atoum/inspections/f38bbe73-1cbe-4bbf-9e01-24ba357c1951) was created.

scrutinizer-notifier commented:
The [inspection](https://scrutinizer-ci.com/g/atoum/atoum/inspections/f38bbe73-1cbe-4bbf-9e01-24ba357c1951) completed: **95 new issues, 27 updated code elements**

Hywan commented:
\o/
